README.md
# MMM-moonraker

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

![screenshot](.github/example.png)

This module will use the Moonraker API to show your current print status on your MagicMirror²! Moonraker is an API that is used by tools like Fluidd or Mainsail to manage your Klipper powered 3D printer with a web interface. The power of the same API is used to display printing related information on your MagicMirror².

## Installation

In your terminal, go to your MagicMirror's Module folder:

```bash
cd ~/MagicMirror/modules
```

Clone this repository:

```bash
git clone https://github.com/btastic/MMM-moonraker
```

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-moonraker',
            position: 'top_right',
            config: {
                endpoint: "",
                updateInterval: 60000,
                showThumbnail: true,
                thumbnailSize: 150,
                hideDataOnStandby: true
            }
        }
    ]
}
```

## Layer count
To see the current layer and the total layer you have to setup your slicer to emit some G-code which will be interpreted by moonraker.
Check the Mainsail documentation on how to add layer information with your slicer: [Layer Count](https://docs.mainsail.xyz/overview/slicer/prusaslicer#layer-count)

## Customization
To change the color of the progress bar, you can override the default colors (green on gray) with css in your custom.css file.

You can also change the border radius of the progress bar, if you don't like rounded corners.

For example:
```css
.MMM-moonraker {
  --progress-color: crimson;
  --progress-background-color: gray;
  --border-radius: 0px;
}
```

## Configuration options

| Option               | Description
|--------------------- |-----------
| `endpoint`           | *Required* <br>Endpoint of the Moonraker API. Usually, this is the same as your Fluidd or Mainsail installation. For example: http://192.168.0.75<br><br>**Type:** `string`(url) 
| `updateInterval`     | *Optional* <br>The time interval between UI updates.<br><br>**Type:** `int`(milliseconds)<br>**Default:** 60000 milliseconds (60 seconds)
| `showThumbnail`      | *Optional* <br>Shows the largest thumbnail embedded in your gcode. Check the Mainsail documentation on how to add thumbnails with your slicer. [How to config your slicer for thumbnails](https://docs.mainsail.xyz/overview/features/thumbnails#how-to-config-your-slicer-for-thumbnails). Make sure to modify your config.js `address` to be something else than `0.0.0.0` or CORS will prevent you from showing the thumbnail. I use `localhost` and it is working fine.<br><br>**Type:** `bool` <br>**Default:** true
| `thumbnailSize`      | *Optional* <br>Changes with thumbnails size (width). Enter a integer value for an absolute value, or a relative size with % (e.g. "50%").<br><br>**Type:** `int\|string` <br>**Default:** 150
| `hideDataOnStandby`  | *Optional* <br>Hides printer info when the printer is on standby.<br><br>**Type:** `bool` <br>**Default:** true

## Available localizations
- English (en)
- German (de)

CHANGELOG.md
# MMM-moonraker Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0]

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-moonraker
 *
 * By Ben Konsemüller
 * MIT Licensed.
 */

const Log = require("logger");
const NodeHelper = require("node_helper");
const fetch = require("fetch");
const moment = require("moment");

module.exports = NodeHelper.create({
  config: {},
  currentFile: null,
  metadata: null,
  async socketNotificationReceived(notification, payload) {
    if (notification === "CONFIG") {
      this.config = payload;

      if (this.fetchTimerId) {
        clearTimeout(this.fetchTimerId);
      }

      await this.fetchData();
    }
  },

  async fetchData() {
    const self = this;

    const printer_status = await this.fetchPrinterStatus();

    if (!printer_status) {
      return;
    }

    if (this.currentFile !== printer_status.print_stats.filename) {
      this.currentFile = printer_status.print_stats.filename;
      this.metadata = await this.fetchMetadata(printer_status);
    }

    const thumbnail = this.metadata?.thumbnails[0].relative_path;

    const eta = await this.calculateEta(printer_status, this.metadata);

    this.sendSocketNotification("PRINTER_STATUS", { printer_status, metadata: this.metadata, eta, thumbnail });

    this.fetchTimerId = setTimeout(async function () {
      await self.fetchData();
    }, this.config.updateInterval);
  },

  async fetchPrinterStatus() {
    const endpoint = this.config.endpoint + "/printer/objects/query?print_stats&virtual_sdcard&extruder&heater_bed";

    try {
      const response = await fetch(endpoint);
      const json = await response.json();

      return json.result.status;
    } catch (error) {
      Log.log(`${this.name} received an error: ${error}`);
      this.sendSocketNotification("HTTP_ERROR", {});

      return null;
    }
  },

  async fetchMetadata(printer_status) {
    if (!printer_status.print_stats.filename) {
      return;
    }

    const endpoint = this.config.endpoint + "/server/files/metadata?filename=" + printer_status.print_stats.filename;

    try {
      const response = await fetch(endpoint);
      const json = await response.json();

      // join up the final thumbnail url for convenience
      if (json.result.thumbnails) {
        for (let index = 0; index < json.result.thumbnails.length; index++) {
          const element = json.result.thumbnails[index];
          element.relative_path = this.config.endpoint + "/server/files/gcodes/" + element.relative_path;
        }

        // sort thumbnails by width so the first one is the larger one
        json.result.thumbnails.sort(function (a, b) { return b.width - a.width });
      }
      
      if (!json.result.thumbnails || json.result.thumbnails.length === 0) {
        json.result.thumbnails = [{ relative_path: "./modules/MMM-moonraker/img/no_thumbnail.png" }];
      }

      return json.result;
    } catch (error) {
      Log.log(`${this.name} received an error: ${error}`);
      this.sendSocketNotification("HTTP_ERROR", {});

      return null;
    }
  },

  calculateEta(printer_status, metadata) {
    if (printer_status.print_stats.state !== 'printing') {
      return null;
    }

    if (metadata && metadata.estimated_time) {
      const progressTime = printer_status.virtual_sdcard.progress * metadata.estimated_time;
      return moment.utc(1000 * (metadata.estimated_time - progressTime)).format('HH[h] mm[m] ss[s]');;
    }

    const total_time = printer_status.print_stats.print_duration / printer_status.virtual_sdcard.progress;
    return moment.utc(1000 * (total_time - printer_status.print_stats.print_duration)).format('HH[h] mm[m] ss[s]');
  },
});


MMM-moonraker.js
/* global Module */

/* Magic Mirror
 * Module: MMM-moonraker
 *
 * By Ben Konsemüller
 * MIT Licensed.
 */

Module.register("MMM-moonraker", {
  defaults: {
    updateInterval: 60000,
    endpoint: "",
    showThumbnail: true,
    thumbnailSize: 150,
    hideDataOnStandby: true
  },

  // Store the data in an object.
  displayData: {
    printer_status: null,
    eta: null,
    thumbnails: [],
  },

  available: true,
  loading: true,

  requiresVersion: "2.1.0", // Required version of MagicMirror

  start: function () {
    Log.info(`Starting module: ${this.name}`);
    this.sendSocketNotification("CONFIG", this.config);
  },

  socketNotificationReceived: function (notification, payload) {
    switch (notification) {
      case "PRINTER_STATUS":
        this.displayData = payload;
        this.available = true;
        this.loading = false;
        break;
      case "HTTP_ERROR":
        this.loading = false;
        this.available = false;
        break;
    }
    this.updateDom(this.config.animationSpeed);
  },

  getTemplate: function () {
    return "templates\\mmm-moonraker.njk";
  },

  getTemplateData() {
    const templateData = {
      loading: this.loading,
      config: this.config,
      data: this.loading ? null : this.displayData.printer_status,
      eta: this.loading ? null : this.displayData.eta,
      thumbnail: this.loading ? null : this.displayData.thumbnail,
      available: this.available,
    };

    return templateData;
  },

  getScripts: function () {
    return [];
  },

  getStyles: function () {
    return [
      "MMM-moonraker.css",
    ];
  },

  getTranslations: function () {
    return {
      en: "translations/en.json",
      de: "translations/de.json",
    };
  },
});
